#!/bin/bash
#$ -S /bin/bash
#$ -cwd
#$ -j y
#$ -m a
#$ -l  h_data=2G,h_rt=24:00:00,h_vmem=12G
#$ -o log.$JOB_ID.log
#$ -t 1-1:1


. /u/local/Modules/default/init/modules.sh

### LIST OF INPUT FILES
input_file=/u/home/c/colinrsm/cfSNV/demo/input_list_getbam_align_after_merge.txt

### PATH TO GATK version 3.8.0 (ABSOLUTE PATH TO .jar FILE)
GATK=/u/home/c/colinrsm/tools/gatk/GenomeAnalysisTK.jar

### PATH TO PICARD TOOLS (ABSOLUTE PATH TO .jar FILE)
picard=/u/home/c/colinrsm/tools/picard/picard.jar

### PATH TO BWA (ABSOLUTE PATH TO EXECUTABLE FILE)
bwa=/u/local/apps/bwa/0.7.17/gcc-4.4.7/bwa

### PATH TO SAMTOOLS (ABSOLUTE PATH TO EXECUTABLE FILE)
samtools=/u/local/apps/samtools/1.9/bin/samtools

### PATH TO JAVA (ABSOLUTE PATH TO EXECUTABLE FILE)
java=/u/local/apps/java/jdk1.8.0_111/bin/java

### PATH TO BEDTOOLS (ABSOLUTE PATH TO EXECUTABLE FILE)
bedtools=/u/local/apps/bedtools/2.26.0/gcc-4.4.7/bin/bedtools

### PATH TO FLASH (ABSOLUTE PATH TO EXECUTABLE FILE)
FLASh=/u/home/c/colinrsm/tools/flash2/flash2


module load samtools

read path fastq1 fastq2 file_id SNP_database reference <<< "`head -$SGE_TASK_ID ${input_file} | tail -1`"

output_dir=${path}

if [ ! -d "$output_dir" ]; then
	mkdir $output_dir
fi


SECONDS=0

cd ${path}

${java} -Xms1g -jar ${picard} CreateSequenceDictionary REFERENCE=${reference} OUTPUT=${reference%.*}.dict
reference_dict=${reference%.*}.dict

${FLASh} -M 20 -x 0.01 -z -q -o ${file_id} ${fastq1} ${fastq2}
${bwa} mem -t 4 ${reference} ${file_id}.notCombined_1.fastq.gz ${file_id}.notCombined_2.fastq.gz | ${samtools} sort -@3 -O BAM -o ${file_id}.notCombined.sorted.bam -
${bwa} mem -t 4 ${reference} ${file_id}.extendedFrags.fastq.gz  | ${samtools} sort -@3 -O BAM -o ${file_id}.extendedFrags.sorted.bam -
${samtools} index ${file_id}.extendedFrags.sorted.bam
${samtools} index ${file_id}.notCombined.sorted.bam

${java} -Xms1G -jar ${picard} MarkDuplicates I=${file_id}.extendedFrags.sorted.bam O=${file_id}.extendedFrags.sorted.rmdup.bam M=${file_id}.extendedFrags.marked_dup_metrics.txt REMOVE_DUPLICATES=true
${java} -Xms1G -jar ${picard} AddOrReplaceReadGroups I=${file_id}.extendedFrags.sorted.rmdup.bam  O=${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam RGID=4 RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=20
rm ${file_id}.extendedFrags.sorted.rmdup.bam ${file_id}.extendedFrags.sorted.bam ${file_id}.extendedFrags.sorted.bam.bai
${java} -Xms1G -jar ${picard} MarkDuplicates I=${file_id}.notCombined.sorted.bam O=${file_id}.notCombined.sorted.rmdup.bam M=${file_id}.notCombined.marked_dup_metrics.txt REMOVE_DUPLICATES=true
${java} -Xms1G -jar ${picard} AddOrReplaceReadGroups I=${file_id}.notCombined.sorted.rmdup.bam  O=${file_id}.notCombined.sorted.rmdup.addreadgroup.bam RGID=4 RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=20
${samtools} index ${file_id}.notCombined.sorted.rmdup.addreadgroup.bam
${samtools} index ${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam
rm ${file_id}.notCombined.sorted.rmdup.bam ${file_id}.notCombined.sorted.bam ${file_id}.notCombined.sorted.bam.bai

${java} -Xms1G -jar ${GATK}  -T RealignerTargetCreator -R ${reference} -I ${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam -o ${file_id}.extendedFrags.forIndelRealigner.intervals
${java} -Xms1G -jar ${GATK} -T IndelRealigner  -R ${reference} -I ${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam  -targetIntervals ${file_id}.extendedFrags.forIndelRealigner.intervals  -o ${file_id}.extendedFrags.realignedBam.bam
rm ${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam ${file_id}.extendedFrags.sorted.rmdup.addreadgroup.bam.bai
${java} -Xms1G -jar ${GATK}  -T RealignerTargetCreator -R ${reference} -I ${file_id}.notCombined.sorted.rmdup.addreadgroup.bam -o ${file_id}.notCombined.forIndelRealigner.intervals
${java} -Xms1G -jar ${GATK} -T IndelRealigner  -R ${reference} -I ${file_id}.notCombined.sorted.rmdup.addreadgroup.bam  -targetIntervals ${file_id}.notCombined.forIndelRealigner.intervals  -o ${file_id}.notCombined.realignedBam.bam
rm ${file_id}.notCombined.sorted.rmdup.addreadgroup.bam ${file_id}.notCombined.sorted.rmdup.addreadgroup.bam.bai

${java} -Xms1G -jar ${GATK} -T BaseRecalibrator -R ${reference} -I ${file_id}.extendedFrags.realignedBam.bam -o ${file_id}.extendedFrags.recal_data.table -knownSites ${SNP_database}
${java} -Xms1G -jar ${GATK} -T PrintReads -R ${reference} -I ${file_id}.extendedFrags.realignedBam.bam -BQSR ${file_id}.extendedFrags.recal_data.table -o ${file_id}.extendedFrags.recal.bam
rm ${file_id}.extendedFrags.realignedBam.bam ${file_id}.extendedFrags.realignedBam.bai
${java} -Xms1G -jar ${GATK} -T BaseRecalibrator -R ${reference} -I ${file_id}.notCombined.realignedBam.bam -o ${file_id}.notCombined.recal_data.table -knownSites ${SNP_database}
${java} -Xms1G -jar ${GATK} -T PrintReads -R ${reference} -I ${file_id}.notCombined.realignedBam.bam -BQSR ${file_id}.notCombined.recal_data.table -o ${file_id}.notCombined.recal.bam
rm ${file_id}.notCombined.realignedBam.bam ${file_id}.notCombined.realignedBam.bai

rm *.intervals *.table *.marked_dup_metrics.txt



duration=${SECONDS}
if (( ${duration} < 1200 )); then
sleep 10m
fi



